# Importing Libraries

In [1]:
pip install geopy

Note: you may need to restart the kernel to use updated packages.


In [3]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: / 
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
                                                             \                                       failed

UnsatisfiableError: The following specifications were found
to be incompatible with the existing python installation in your environment:

Specifications:

  - cffi -> python[version='2.7.*|3.5.*|3.6.*|3.6.12|>=3.6,<3.7.0a0|>=3.7,<3.8.0a0|>=3.9,<3.10.0a0|>=3.8,<3.9.0a0|3.6.9|3.6.9|3.6.9|>=2.7,<2.8.0a0|3.6.9|>=3.5,<3.6.0a0|3.4.*',build='2_73_pypy|4_73_pypy|3_73_pypy|1_73_pypy|0_73_pypy']
  - geopy -> python[version='2.7.*|3.4.*|3.5.*|3.6.*']
  - rsa -> python[version='2.7.*|3.4.

Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: \ 
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
                                                                                                    -failed

UnsatisfiableError: The following specifications were found
to be incompatible with the existing python installation in your environment:

Specifications:

  - cffi -> python[version='2.7.*|3.5.*|3.6.*|3.6.12|>=3.6,<3.7.0a0|>=3.7,<3.8.0a0|>=3.9,<3.10.0a0|>=3.8,<3.9.0a0|3.6.9|3.6.9|3.6.9|>=2.7,<2.8.0a0|3.6.9|>=3.5,<3.6.0a0|3.4.*',build='1_73_pypy|2_73_pypy|4_73_pypy|3_73_pypy|0_73_pypy']
  - rsa -> python[version='2.7.*|3.4.*|3.5.*|3.6.*']

Your python: python=3.7

If python is on the left-most side of the chain, that's the version you've asked for.
When python appears to the right, t

Libraries imported.


In [2]:
pip install folium

Note: you may need to restart the kernel to use updated packages.


# Downloading Data in Dataframe

In [29]:
from bs4 import BeautifulSoup

In [30]:
url= 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

source= requests.get(url).text
soup= BeautifulSoup(source)
table_data=soup.find('div',
class_='mw-parser-output')
table=table_data.table.tbody

columns= ['Postal Code', 'Borough', 'Neighborhood']
data=dict({key:[]*len(columns) for key in columns})

for row in table.find_all('tr'):
    for i, column in zip(row.find_all('td'),columns):
        i=i.text
        i=i.replace('\n',
'')
        data[column].append(i)
        
df= pd.DataFrame.from_dict(data=data)[columns]
print(df.shape)
df.head()

(180, 3)


Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

# Dropping Not Assigned Boroughs

In [32]:
df= df[df['Borough'] != 'Not assigned'].reset_index(drop = True)
print('After dropping rows where borough is "Not Assigned", Shape is: ',df.shape)
print('Number of rows where Neighborhood is "Not Assigned" but borough has value: ',
     df[df['Neighborhood'] == 'Not assigned'].shape[0])

After dropping rows where borough is "Not Assigned", Shape is:  (103, 3)
Number of rows where Neighborhood is "Not Assigned" but borough has value:  0


In [33]:
p, b, n =[], [], []

for postcode, borough, neigh in zip(df['Postal Code'], df['Borough'], df['Neighborhood']):
    p.append(postcode)
    b.append(borough)
    if neigh == 'Not assigned':
        n.append(borough)
    else:
        n.append(neigh)

df=pd.DataFrame({'Postal Code':p,'Borough':b, 'Neighborhood':n})[columns]
print(df.shape)
df.head()

(103, 3)


Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

# <font size="4"> 77 rows have been dropped because there was no assigned boroughs for them.</font>

# <font size="5"> Original Data had 180 rows and now the data is having 103 rows.</font>

In [35]:
postcodes=df['Postal Code'].values
boroughs =df['Borough'].values
neighs=df['Neighborhood'].values

dic=dict({(key1, key2): [] for key1, key2 in zip(postcodes, boroughs)})
print('Number of keys in the dictionary are: ',len(dic.keys()))

for postcode, borough, neigh in zip(postcodes, boroughs, neighs):
    key=(postcode, borough)
    dic[key].append(neigh)

df=pd.DataFrame(columns=['Postal Code','Borough','Neighborhood'])
for key, value in dic.items():
    postcode, borough, neig=key[0],key[1], value
    neig=', '.join(neig)
    df=df.append({'Postal Code': postcode,
                 'Borough': borough,
                 'Neighborhood': neig}, ignore_index=True)

print('Shape of final data is: ',df.shape)
df.head(10)

Number of keys in the dictionary are:  103
Shape of final data is:  (103, 3)


Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village
6         M1B       Scarborough                               Malvern, Rouge
7         M3B        North York                                    Don Mills
8         M4B         East York              Parkview Hill, Woodbine Gardens
9         M5B  Downtown Toronto                     Garden District, Ryerson

# Latitude and Longitude of Each Neighbourhood

In [18]:
pip install geocoder

     |████████████████████████████████| 98 kB 8.6 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [19]:
import geocoder # import geocoder

In [21]:
pip install pgeocode

Note: you may need to restart the kernel to use updated packages.


# Downlaoding the GeoSpatial Data by downloading coordinates of postal codes

In [36]:
!wget http://cocl.us/Geospatial_data

--2020-12-19 14:25:29--  http://cocl.us/Geospatial_data
Resolving cocl.us (cocl.us)... 169.63.96.176, 169.63.96.194
Connecting to cocl.us (cocl.us)|169.63.96.176|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://cocl.us/Geospatial_data [following]
--2020-12-19 14:25:29--  https://cocl.us/Geospatial_data
Connecting to cocl.us (cocl.us)|169.63.96.176|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2020-12-19 14:25:30--  https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv
Resolving ibm.box.com (ibm.box.com)... 185.235.236.197
Connecting to ibm.box.com (ibm.box.com)|185.235.236.197|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /public/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2020-12-19 14:25:30--  https://ibm.box.com/public/static/9afzr8

In [37]:
try:
    df['Latitude']= lats
    df['Longitude']= lons
except:
    latlon=pd.read_csv('Geospatial_data')
    df=pd.merge(df, latlon, how='inner', on= 'Postal Code')

print(df.shape)
df.head(10)

(103, 5)


Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village   
6         M1B       Scarborough                               Malvern, Rouge   
7         M3B        North York                                    Don Mills   
8         M4B         East York              Parkview Hill, Woodbine Gardens   
9         M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
7  43.745906 -79.352188  
8  43.706397 -79.309937  
9  43.657162 -79.378937

# Exploring and Clustering Toronto Neighborhoods

In [38]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


In [39]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [40]:
print('Total number of Borough= ', len(df['Borough'].unique()))

Total number of Borough=  10


# Exploring one borough for reference at the start and then we will check other neighbourhoods as well

In [42]:
downtown_toronto=df[df['Borough']=='Downtown Toronto'].reset_index(drop=True)
print(downtown_toronto.shape)
downtown_toronto.head()

(19, 5)


Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

# Visualized Map of Downtown Toronto

In [46]:
address = 'Downtown Toronto, Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

# create map of Downtown Toronto using latitude and longitude values
map_downtown = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(downtown_toronto['Latitude'], downtown_toronto['Longitude'], downtown_toronto['Borough'],downtown_toronto['Neighborhood']):
    label='{},{}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown)  
    
map_downtown

# Exploring Venues using Foursquare API

In [47]:
lat= downtown_toronto.loc[0,'Latitude']
lon= downtown_toronto.loc[0,'Longitude']
neighborhood_name= downtown_toronto.loc[0,'Neighborhood']

print('Latitude and Longitude of the neighborhood as {} are {}, {} '.format(neighborhood_name, lat,lon))

Latitude and Longitude of the neighborhood as Regent Park, Harbourfront are 43.6542599, -79.3606359 


In [50]:
CLIENT_ID = 'C11DJYAJWWUVQFE4UHBQ5A24ZKZBHVMHYJQLJ4MOK5H1YL1J' # your Foursquare ID
CLIENT_SECRET = '3IPTYS4WO4AAGN1FLIDQVL5O3SLSOS4TSJTLBWDEC4Y22OQN' # your Foursquare Secret
VERSION = '20201219' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

In [52]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 1000 # define radius
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    lat, 
    lon, 
    radius, 
    LIMIT)

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5fde1b6dd358eb7aca789e69'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 109,
  'suggestedBounds': {'ne': {'lat': 43.66325990900001,
    'lng': -79.3482199002972},
   'sw': {'lat': 43.64525989099999, 'lng': -79.37305189970282}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label':

In [53]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [54]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  app.launch_new_instance()


name           categories        lat        lng
0                  Roselle Desserts               Bakery  43.653447 -79.362017
1                     Tandem Coffee          Coffee Shop  43.653559 -79.361809
2            Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3                    Impact Kitchen           Restaurant  43.656369 -79.356980
4  The Distillery Historic District        Historic Site  43.650244 -79.359323

# Venues and Details for All Neighborhoods of Toronto

In [55]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [56]:
downtown_venues = getNearbyVenues(names=downtown_toronto['Neighborhood'],
                                   latitudes=downtown_toronto['Latitude'],
                                   longitudes=downtown_toronto['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


In [57]:
print(downtown_venues.shape)
downtown_venues.head()

(1238, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

In [59]:
downtown_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            61   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     65   
Christie                                                               16   
Church and Wellesley                                                   80   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              58   
Queen's Park, Ontario Provincial Government                            32   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                               95   
Rosedale                                                                4   
St. James Town                                                         83   
St. James Town, Cabbagetown                                            50   
Stn A PO Boxes                                                         98   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             61   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      65   
Christie                                                                16   
Church and Wellesley                                                    80   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               58   
Queen's Park, Ontario Provincial Government                             32   
Regent Park, Harbourfront                                               45   
Richmond, Adelaide, King                                                95   
Rosedale                                                                 4   
St. James Town                                                          83   
St. James Town, Cabbagetown                                             50   
Stn A PO Boxes                                                          98   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          34   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            61              61   
CN Tower, King and Spadina, Railway Lands, Harb...     17              17   
Central Bay Street                                     65              65   
Christie                                               16              16   
Church and Wellesley                                   80              80   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

In [60]:
print('There are {} unique categories.'.format(len(downtown_venues['Venue Category'].unique())))

There are 211 unique categories.


# Analyzing each neighborhood

In [61]:
# one hot encoding
downtown_onehot = pd.get_dummies(downtown_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
downtown_onehot['Neighborhood'] = downtown_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [downtown_onehot.columns[-1]] + list(downtown_onehot.columns[:-1])
downtown_onehot = downtown_onehot[fixed_columns]

downtown_onehot.head()

Yoga Studio  Adult Boutique  Airport  Airport Food Court  Airport Gate  \
0            0               0        0                   0             0   
1            0               0        0                   0             0   
2            0               0        0                   0             0   
3            0               0        0                   0             0   
4            0               0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  \
0                 0                   0          0           0           0   
1                 0                   0          0           0           0   
2                 0                   0          0           0           0   
3                 0                   0          0           0           0   
4                 0                   0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       1     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Butcher  \
0        0                0         0             0              0        0   
1        0                0         0             0              0        0   
2        0                0         0             0              0        0   
3        0                0         0             0              0        0   
4        0                0         0             0              0        0   

   Café  Candy Store  Caribbean Restaurant  Cheese Shop  Chinese Restaurant  \
0     0            0                     0            0                   0   
1     0            0                     0            0                   0   
2     0            0                     0            0                   0   
3     0            0                     0 

In [63]:
downtown_onehot.shape

(1238, 211)

In [64]:
downtown_grouped = downtown_onehot.groupby('Neighborhood').mean().reset_index()
downtown_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
2                                  Central Bay Street     0.015385   
3                                            Christie     0.000000   
4                                Church and Wellesley     0.025000   
5                      Commerce Court, Victoria Hotel     0.000000   
6              First Canadian Place, Underground city     0.000000   
7                            Garden District, Ryerson     0.000000   
8   Harbourfront East, Union Station, Toronto Islands     0.000000   
9           Kensington Market, Chinatown, Grange Park     0.000000   
10        Queen's Park, Ontario Provincial Government     0.031250   
11                          Regent Park, Harbourfront     0.022222   
12                           Richmond, Adelaide, King     0.000000   
13                                           Rosedale     0.000000   
14                                     St. James Town     0.000000   
15                        St. James Town, Cabbagetown     0.000000   
16                                     Stn A PO Boxes     0.010204   
17           Toronto Dominion Centre, Design Exchange     0.000000   
18                     University of Toronto, Harbord     0.029412   

    Adult Boutique   Airport  Airport Food Court  Airport Gate  \
0           0.0000  0.000000            0.000000      0.000000   
1           0.0000  0.058824            0.058824      0.058824   
2           0.0000  0.000000            0.000000      0.000000   
3           0.0000  0.000000            0.000000      0.000000   
4           0.0125  0.000000            0.000000      0.000000   
5           0.0000  0.000000            0.000000      0.000000   
6           0.0000  0.000000            0.000000      0.000000   
7           0.0000  0.000000            0.000000      0.000000   
8           0.0000  0.000000            0.000000      0.000000   
9           0.0000  0.000000            0.000000      0.000000   
10          0.0000  0.000000            0.000000      0.000000   
11          0.0000  0.000000            0.000000      0.000000   
12          0.0000  0.000000            0.000000      0.000000   
13          0.0000  0.000000            0.000000      0.000000   
14          0.0000  0.000000            0.000000      0.000000   
15          0.0000  0.000000            0.000000      0.000000   
16          0.0000  0.000000            0.000000      0.000000   
17          0.0000  0.000000            0.000000      0.000000   
18          0.0000  0.000000            0.000000      0.000000   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0         0.000000         0.000000          0.000000             0.000000   
1         0.117647         0.117647          0.117647             0.000000   
2         0.000000         0.000000          0.000000             0.000000   
3         0.000000         0.000000          0.000000             0.000000   
4         0.000000         0.000000          0.000000             0.012500   
5         0.000000         0.000000          0.000000             0.030000   
6         0.000000         0.000000          0.000000             0.030000   
7         0.000000         0.000000          0.000000             0.000000   
8         0.000000         0.000000          0.000000             0.000000   
9         0.000000         0.000000          0.000000             0.000000   
10        0.000000         0.000000          0.000000             0.000000   
11        0.000000         0.000000          0.000000             0.000000   
12        0.000000         0.000000          0.000000             0.021053   
13        0.000000         0.000000          0.000000             0.000000   
14        0.000000         0.000000          0.000000             0.036145   
15        0.000000         0.000000          0.000000             0.000

In [66]:
downtown_grouped.shape

(19, 211)

# 5 Most Common Venues of Each Neighborhood

In [67]:
num_top_venues = 5

for hood in downtown_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = downtown_grouped[downtown_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
            venue  freq
0     Coffee Shop  0.10
1          Bakery  0.05
2    Cocktail Bar  0.05
3      Restaurant  0.03
4  Farmers Market  0.03


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.12
1   Airport Service  0.12
2  Airport Terminal  0.12
3     Boat or Ferry  0.06
4             Plane  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.17
1                Café  0.06
2      Sandwich Place  0.06
3  Italian Restaurant  0.05
4     Thai Restaurant  0.03


----Christie----
           venue  freq
0  Grocery Store  0.25
1           Café  0.19
2           Park  0.12
3      Nightclub  0.06
4     Restaurant  0.06


----Church and Wellesley----
                 venue  freq
0          Coffee Shop  0.06
1     Sushi Restaurant  0.06
2  Japanese Restaurant  0.06
3           Restaurant  0.04
4              Gay Bar  0.04


--

# Checking 10 Most Common Venue of Each Neighborhood

In [68]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [69]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = downtown_grouped['Neighborhood']

for ind in np.arange(downtown_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(downtown_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery          Cocktail Bar            Restaurant   
1       Airport Service      Airport Terminal      Sculpture Garden   
2                  Café        Sandwich Place    Italian Restaurant   
3                  Café                  Park    Athletics & Sports   
4   Japanese Restaurant      Sushi Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Farmers Market    Seafood Restaurant           Cheese Shop   
1       Harbor / Marina         Boat or Ferry   Rental Car Location   
2                  Bank           Salad Place       Bubble Tea Shop   
3    Italian Restaurant             Nightclub           Candy Store   
4               Gay Bar           Yoga Studio                   Pub   

  8th Most Common Venue     9th Most Common Venue      10th Most Common Venue  
0              Pharmacy                  Beer Bar                        Park  
1                   Bar                     Plane                 Coffee Shop  
2          Burger Joint           Thai Restaurant  Modern European Restaurant  
3            Restaurant                Baby Store                 Coffee Shop  
4           Men's Store  Mediterranean Restaurant                       Hotel

# Clustering Neighbours

In [76]:
# set number of clusters
kclusters = 5

downtown_grouped_clustering = downtown_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(downtown_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 0, 1, 2, 1, 1, 1, 1, 1, 4], dtype=int32)

In [78]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels_1', kmeans.labels_)

downtown_merged = downtown_toronto

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
downtown_merged = downtown_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

downtown_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  Cluster Labels_1  Cluster Labels  \
0  43.654260 -79.360636                 1               1   
1  43.662301 -79.389494                 1               1   
2  43.657162 -79.378937                 1               1   
3  43.651494 -79.375418                 1               1   
4  43.644771 -79.373306                 1               1   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop                  Park                  Café   
1           Coffee Shop     College Cafeteria      Sushi Restaurant   
2        Clothing Store           Coffee Shop                  Café   
3           Coffee Shop                  Café   American Restaurant   
4           Coffee Shop                Bakery          Cocktail Bar   

  4th Most Common Venue 5th Most Common Venue      6th Most Common Venue  \
0                Bakery                   Pub             Breakfast Spot   
1           Yoga Studio                 Diner                        Bar   
2        Cosmetics Shop       Bubble Tea Shop  Middle Eastern Restaurant   
3             Gastropub        Cosmetics Shop               Cocktail Bar   
4            Restaurant        Farmers Market         Seafood Restaurant   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0               Theater           Yoga Studio        Farmers Market   
1        Sandwich Place    Italian Restaurant   Japanese Restaurant   
2   Japanese Restaurant           Pizza Place                 Diner   
3        Clothing Store                   Gym        Farmers Market   
4           Cheese Shop              Pharmacy              Beer Bar   

  10th Most Common Venue  
0             Shoe Store  
1    Distribution Center  
2              Bookstore  
3       Department Store  
4                   Park

In [85]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(downtown_merged['Latitude'], downtown_merged['Longitude'], downtown_merged['Neighborhood'], downtown_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examining Clusters

### Cluster 1

In [86]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 0, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels_1  Cluster Labels 1st Most Common Venue  \
13  Downtown Toronto                 0               0        Airport Lounge   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13       Airport Service      Airport Terminal      Sculpture Garden   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13       Harbor / Marina         Boat or Ferry   Rental Car Location   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13                   Bar                 Plane            Coffee Shop

### Cluster 2

In [87]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 1, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels_1  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto                 1               1           Coffee Shop   
1   Downtown Toronto                 1               1           Coffee Shop   
2   Downtown Toronto                 1               1        Clothing Store   
3   Downtown Toronto                 1               1           Coffee Shop   
4   Downtown Toronto                 1               1           Coffee Shop   
5   Downtown Toronto                 1               1           Coffee Shop   
7   Downtown Toronto                 1               1           Coffee Shop   
8   Downtown Toronto                 1               1           Coffee Shop   
9   Downtown Toronto                 1               1           Coffee Shop   
10  Downtown Toronto                 1               1           Coffee Shop   
15  Downtown Toronto                 1               1           Coffee Shop   
16  Downtown Toronto                 1               1           Coffee Shop   
17  Downtown Toronto                 1               1           Coffee Shop   
18  Downtown Toronto                 1               1           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Park                  Café                Bakery   
1      College Cafeteria      Sushi Restaurant           Yoga Studio   
2            Coffee Shop                  Café        Cosmetics Shop   
3                   Café   American Restaurant             Gastropub   
4                 Bakery          Cocktail Bar            Restaurant   
5                   Café        Sandwich Place    Italian Restaurant   
7                   Café            Restaurant         Deli / Bodega   
8               Aquarium                 Hotel                  Café   
9                  Hotel                  Café            Restaurant   
10            Restaurant                  Café                 Hotel   
15    Italian Restaurant    Seafood Restaurant          Cocktail Bar   
16                  Café    Chinese Restaurant                Market   
17                  Café                 Hotel            Restaurant   
18   Japanese Restaurant      Sushi Restaurant            Restaurant   

   5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0                    Pub             Breakfast Spot               Theater   
1                  Diner                        Bar        Sandwich Place   
2        Bubble Tea Shop  Middle Eastern Restaurant   Japanese Restaurant   
3         Cosmetics Shop               Cocktail Bar        Clothing Store   
4         Farmers Market         Seafood Restaurant           Cheese Shop   
5                   Bank                Salad Place       Bubble Tea Shop   
7        Thai Restaurant                        Gym        Clothing Store   
8    Fried Chicken Joint                 Restaurant        Scenic Lookout   
9    Japanese Restaurant         Italian Restaurant    Seafood Restaurant   
10    Italian Restaurant                        Gym   American Restaurant   
15   Japanese Restaurant                     Bakery            Restaurant   
16            Restaurant                        Pub           Pizza Place   
17                   Gym        Japanese Restaurant   American Restaurant   
18               Gay Bar                Yoga Studio                   Pub   

   8th Most Common Venue     9th Most Common Venue      10th Most Common Venue  
0            Yoga Studio            Farmers Market                  Shoe Store  
1     Italian Restaurant       Japanese Restaurant         Distribution Center  
2            Pizza Place                     Diner                   Bookstore  
3                    Gym            Farmers Market            Department Store  
4               Pharmacy                  Beer Bar                        Park  
5           Burger Joint           Thai Restaurant  Modern European Restaurant  
7       

### Cluster 3

In [88]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 2, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels_1  Cluster Labels 1st Most Common Venue  \
6  Downtown Toronto                 2               2         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                  Café                  Park    Athletics & Sports   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6    Italian Restaurant             Nightclub           Candy Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6            Restaurant            Baby Store            Coffee Shop

### Cluster 4

In [89]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 3, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels_1  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto                 3               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14                 Trail            Playground              Wine Bar   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14          Cupcake Shop            Donut Shop      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14               Dog Run   Distribution Center         Discount Store

### Cluster 5

In [90]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 4, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels_1  Cluster Labels 1st Most Common Venue  \
11  Downtown Toronto                 4               4                  Café   
12  Downtown Toronto                 4               4                  Café   

   2nd Most Common Venue  3rd Most Common Venue  \
11                   Bar     Italian Restaurant   
12           Coffee Shop  Vietnamese Restaurant   

            4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
11            Japanese Restaurant             Bookstore                Bakery   
12  Vegetarian / Vegan Restaurant    Mexican Restaurant         Grocery Store   

   7th Most Common Venue    8th Most Common Venue 9th Most Common Venue  \
11           Yoga Studio               Restaurant              Beer Bar   
12          Dessert Shop  Comfort Food Restaurant  Caribbean Restaurant   

   10th Most Common Venue  
11             Beer Store  
12            Gaming Cafe